File: ttnn/ttnn/operations/complex_unary_backward.py
# SPDX-FileCopyrightText: © 2023 Tenstorrent USA, Inc.

# SPDX-License-Identifier: Apache-2.0

import ttnn
from ttnn.operations.golden_common import (
    golden_compute_gradients,
    golden_pack_complex_gradient,
    golden_prepare_grad_inputs,
)


def _complex_unary_backward_reference(forward, grad_tensor, input_tensor):
    (input_tensor,) = golden_prepare_grad_inputs(input_tensor)
    output = forward(input_tensor)
    (input_gradient,) = golden_compute_gradients(output, (input_tensor,), grad_tensor)
    return [golden_pack_complex_gradient(input_gradient)]


def _make_complex_unary_backward(forward):
    def golden_function(grad_tensor, input_tensor, *args, **kwargs):
        import torch

        forward_function = getattr(torch, forward) if isinstance(forward, str) else forward
        return _complex_unary_backward_reference(forward_function, grad_tensor, input_tensor)

    return golden_function


def _polar_forward(input_tensor):
    import torch

    return torch.polar(input_tensor.real, input_tensor.imag)


ttnn.attach_golden_function(ttnn.real_bw, golden_function=_make_complex_unary_backward("real"))
ttnn.attach_golden_function(ttnn.imag_bw, golden_function=_make_complex_unary_backward("imag"))
ttnn.attach_golden_function(ttnn.angle_bw, golden_function=_make_complex_unary_backward("angle"))
ttnn.attach_golden_function(ttnn.conj_bw, golden_function=_make_complex_unary_backward("conj"))
ttnn.attach_golden_function(ttnn.polar_bw, golden_function=_make_complex_unary_backward(_polar_forward))

__all__ = []

File: ttnn/cpp/ttnn/operations/eltwise/complex_unary_backward/complex_unary_backward.cpp
// SPDX-FileCopyrightText: © 2023 Tenstorrent USA, Inc.
//
// SPDX-License-Identifier: Apache-2.0

#include "complex_unary_backward.hpp"

#include "ttnn/operations/eltwise/complex_unary_backward/device/complex_unary_backward_op.hpp"
#include "ttnn/graph/composite_trace.hpp"

namespace ttnn {

std::vector<ComplexTensor> polar_bw(
    const ComplexTensor& grad_tensor, const ComplexTensor& input_tensor, const MemoryConfig& memory_config) {
    TT_OP_SCOPE("ttnn::polar_bw");
    return operations::complex_unary_backward::_polar_bw(grad_tensor, input_tensor, memory_config);
}

std::vector<ComplexTensor> conj_bw(
    const ComplexTensor& grad_tensor, const ComplexTensor& input_tensor, const MemoryConfig& memory_config) {
    TT_OP_SCOPE("ttnn::conj_bw");
    return operations::complex_unary_backward::_conj_bw(grad_tensor, input_tensor, memory_config);
}

std::vector<ComplexTensor> imag_bw(
    const Tensor& grad_tensor, const ComplexTensor& input_tensor, const MemoryConfig& memory_config) {
    TT_OP_SCOPE("ttnn::imag_bw");
    return operations::complex_unary_backward::_imag_bw(grad_tensor, input_tensor, memory_config);
}

std::vector<ComplexTensor> real_bw(
    const Tensor& grad_tensor, const ComplexTensor& input_tensor, const MemoryConfig& memory_config) {
    TT_OP_SCOPE("ttnn::real_bw");
    return operations::complex_unary_backward::_real_bw(grad_tensor, input_tensor, memory_config);
}

std::vector<ComplexTensor> angle_bw(
    const Tensor& grad_tensor, const ComplexTensor& input_tensor, const MemoryConfig& memory_config) {
    TT_OP_SCOPE("ttnn::angle_bw");
    return operations::complex_unary_backward::_angle_bw(grad_tensor, input_tensor, memory_config);
}

}  // namespace ttnn
